taar/recommenders/recommendation_manager.py

```diff
@@ -1,54 +1,62 @@
 import logging
 from .collaborative_recommender import CollaborativeRecommender
 from .locale_recommender import LocaleRecommender
-from .empty_recommender import EmptyRecommender
 from ..profile_fetcher import ProfileFetcher
 
 
 logger = logging.getLogger(__name__)
 
 
-class RecommendationManager:
+class RecommendationManager(object):
+    """A class representing a collection of recommenders."""
 
     def __init__(self, profile_fetcher=None, recommenders=None):
-        # Instantiate the object to get the client info.
+        """Initialize the user profile fetcher and the recommenders.
+
+        Note: The order of the recommenders matters!
+        """
         if profile_fetcher is None:
             logger.info("Initializing profile_fetcher")
             self.profile_fetcher = ProfileFetcher()
         else:
             self.profile_fetcher = profile_fetcher
 
-        # Init the recommenders. Note: ORDER MATTERS!
-        # The recommendation logic will go through each recommender and pick the
-        # first one that "can_recommend".
         if not recommenders:
             logger.info("Initializing recommenders")
             self.recommenders = (
-                CollaborativeRecommender(),
-                LocaleRecommender(),
-                EmptyRecommender()
+                CollaborativeRecommender,
+                LocaleRecommender
             )
         else:
             self.recommenders = recommenders
 
+
     def recommend(self, client_id, limit):
+        """Return recommendations for the given client.
+
+        The recommendation logic will go through each recommender and pick the
+        first one that "can_recommend".
+        """
         # Get the info for the requested client id.
         client_info = self.profile_fetcher.get(client_id)
         if client_info is None:
             return []
 
         # Compute the recommendation.
-        for r in self.recommenders:
-            if r.can_recommend(client_info):
-                recommender_name = r.__class__.__name__
+        for recommender_class in self.recommenders:
+            recommender = recommender_class()
```

Comment:
Mh, wouldn't this mean that, for example, the collaborative recommender module would build an item matrix each time this gets called? (i.e. `_build_model()`).

Do we gain anything in instantiating it each time?

Reply:
Based on the taar-api use case, yes. Before this patch, every recommender is instantiated when the manager init method. Even if it's never going to be used. My patch delays the instantiation to when it's necessary, but it does it on every call to `recommend`. Taar api instantiate a new RecommendationManager on every request, and that's why I didn't think that was a problem. Do we have a use case that gets negatively affected by this change?

Reply:
Mh, I thought the `RecommendationManager` would just be instantiated once by the taar-api. Otherwise, at the very least, we always get the disk I/O hit for loading the cache, right?

Reply:
We discussed this off line, I'm dropping this commit entirely.